File: C/4/12.c
#include <stdio.h>

#define MAX 11
#define MAX_STR 99999

/**
 * @brief Converts an integer into a string recursively.
 *
 * @param n     Number to be converted
 * @param s     Array of characters to be filled
 *
 * @return int  Number of characters inserted into the string
 */
int itoa(int n, char s[])
{
    int pos = 0;

    if (n / 10) {
        pos = itoa(n / 10, s);
    } else {
        if (n < 0) {
            s[pos++] = '-'; /* We add a sign at the first position, in case its a negative number */
        }
    }

    if (n < 0) {
        n = -n;
    }

    s[pos++] = (n % 10) + '0';
    s[pos] = '\0';

    return pos;
}

main()
{
    char result[MAX_STR] = "";

    int i = 0;
    int numbers[] = {
        23123,
        -53352,
        24,
        12,
        0,
        -516361361,
        123632,
        236,
        352,
        1412,
        -12414
    };

    for (i = 0; i < MAX; i++) {
        itoa(numbers[i], result);
        printf("\t[*] %d => \"%s\"\n", numbers[i], result);
    }
}

File: rustlang/4/12.rs
/// @brief Converts an integer into a string recursively.
///
/// @param n     Number to be converted
/// @param s     Array of characters to be filled
fn itoa(mut n: i32, s: &mut String)
{
    if (n / 10) != 0 {
        itoa(n / 10, s);
    } else {
        if n < 0 {
            s.push('-'); // We add a sign at the first position, in case its a negative number
        }
    }

    if n < 0 {
        n = -n;
    }

    s.push(std::char::from_u32((n as u32 % 10) + ('0' as u32)).unwrap_or('#'));
}

fn main()
{
    let mut result: String = String::new();

    let numbers: Vec<i32> = vec![
        23123,
        -53352,
        24,
        12,
        0,
        -516361361,
        123632,
        236,
        352,
        1412,
        -12414
    ];

    for _number in numbers {
        result.clear();

        itoa(_number, &mut result);
        print!("\t[*] {} => \"{}\"\n", _number, result);
    }
}
